pkg/p2p/wire/message/message.go

```diff
@@ -0,0 +1,237 @@
+
+func (m simple) SetPayload(payload interface{}) {
+	m.payload = payload
+}
+
+func (m simple) Equal(other Message) bool {
+	msg, ok := other.(simple)
+	return ok && bytes.Equal(msg.marshalled.Bytes(), msg.marshalled.Bytes())
+}
+
+// New creates a new Message
+func New(t topics.Topic, payload interface{}) Message {
+	return &simple{category: t, payload: payload}
+}
+
+func newMsg(t topics.Topic) *simple {
+	return &simple{category: t}
+}
+
+func (m *simple) initPayloadBuffer(b bytes.Buffer) {
+	if m.marshalled == nil {
+		m.marshalled = bytes.NewBuffer(b.Bytes())
+	}
+}
+
+// Unmarshal mutates the buffer by extracting the topic. It create the Message
+// by setting the topic and unmarshalling the payload into the proper structure
+// It also caches the serialized form within the message
+func Unmarshal(b *bytes.Buffer) (Message, error) {
+	var err error
+	topic, err := topics.Extract(b)
+	if err != nil {
+		return nil, err
+	}
+
+	msg := newMsg(topic)
+	msg.initPayloadBuffer(*b)
+	switch topic {
+	case topics.Tx:
+		err = UnmarshalTxMessage(b, msg)
+	case topics.Candidate:
+		err = UnmarshalCandidateMessage(b, msg)
+	case topics.Score:
+		err = UnmarshalScoreMessage(b, msg)
+	case topics.Reduction:
+		err = UnmarshalReductionMessage(b, msg)
+	case topics.Agreement:
+		err = UnmarshalAgreementMessage(b, msg)
+		//case RoundResults:
+		//err = unmarshalRoundResultMessage(b, msg)
+	}
+
+	if err != nil {
+		return nil, err
+	}
+
+	return msg, nil
+}
+
+// Marshal a Message into a buffer. The buffer *does* include its Category (so
+// if this is undesired, the client of this call needs to explicitly call
+// topics.Extract on the resulting buffer
+// TODO: interface - once the Gossip preprocessing is removed from the
+// Coordinator, there won't be a need for marshalBuffer
+func Marshal(s Message) (bytes.Buffer, error) {
+	var buf bytes.Buffer
+	// if it is a simple message, first we check if this message carries a
+	// cache of its marshalled form first
+	m, ok := s.(simple)
+
+	if !ok { // it is not a cacheable message. We simply marshal without caring for any optimization
+		if err := marshal(s, &buf); err != nil {
+			return bytes.Buffer{}, err
+		}
+		return buf, nil
+	}
+
+	// it is a simple message
+	if m.marshalled != nil {
+		// the message has cached its marshalled form
+		b := *m.marshalled
+		return b, nil
+	}
+
+	// this message has never been marshalled before
+	if err := marshal(m, &buf); err != nil {
+		return bytes.Buffer{}, err
+	}
+
+	m.initPayloadBuffer(buf)
+	return buf, nil
+}
+
+// marshal forces the serialization of a Message regardless of its cache
+func marshal(s Message, b *bytes.Buffer) error {
+	payload := s.Payload()
+	if payload == nil {
+		*b = s.Category().ToBuffer()
+		return nil
+	}
+
+	switch payload.(type) {
+	case bytes.Buffer:
+		*b = payload.(bytes.Buffer)
+
+	default:
+		if err := marshalMessage(s.Category(), payload, b); err != nil {
+			return err
+		}
+	}
+
+	topics.Prepend(b, s.Category())
+	return nil
```

Comment:
nitpick: `return topics.Prepend(b, s.Category())`

don't wanna drop that error value